File: 19. Remove duplicates unsorted.py
class Node:
    def __init__(self,data): self.data=data; self.next=None

def remove_dup_unsorted(head):
    seen=set()
    prev=None
    cur=head
    while cur:
        if cur.data in seen:
            prev.next=cur.next
        else:
            seen.add(cur.data)
            prev=cur
        cur=cur.next
    return head

def print_list(h):
    cur=h; out=[]
    while cur:
        out.append(str(cur.data)); cur=cur.next
    print(' -> '.join(out))

if __name__=='__main__':
    head=Node(1); head.next=Node(2); head.next.next=Node(2); head.next.next.next=Node(3); head.next.next.next.next=Node(1)
    head=remove_dup_unsorted(head)
    print_list(head)

File: 19. Remove duplicates unsorted.cpp
#include <iostream>
#include <unordered_set>
using namespace std; struct Node{int data; Node* next; Node(int d):data(d),next(nullptr){}
};
void removeDup(Node* head){ if(!head) return; unordered_set<int> s; Node* cur=head; Node* prev=nullptr; while(cur){ if(s.count(cur->data)){ prev->next=cur->next; delete cur; cur=prev->next; } else { s.insert(cur->data); prev=cur; cur=cur->next; } } }
void print(Node* h){ while(h){ cout<<h->data; if(h->next) cout<<" -> "; h=h->next;} cout<<endl; }
int main(){ Node* head=new Node(1); head->next=new Node(2); head->next->next=new Node(2); head->next->next->next=new Node(3); head->next->next->next->next=new Node(1); removeDup(head); print(head); return 0; }
